File: Applied Course/4.Problem Solving/5.Problems on Trees/33.Validate Binary Search Tree.cpp
#include<bits/stdc++.h>
using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBSTHelper(TreeNode* root, long min, long max)
    {
        if(!root)
            return true;
        if(root->val>min && root->val<max)
        {
            bool temp1=isValidBSTHelper(root->left,min,root->val);
            bool temp2=isValidBSTHelper(root->right,root->val,max);
            return temp1 && temp2;
        }
        return false;
    }
    bool isValidBST(TreeNode* root) {
        return isValidBSTHelper(root,LONG_MIN,LONG_MAX);
    }
};
int main()
{
    // https://leetcode.com/problems/validate-binary-search-tree/
    return 0;
}

File: Applied Course/4.Problem Solving/5.Problems on Trees/33.Validate Binary Search Tree.py
# https://leetcode.com/problems/validate-binary-search-tree/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Logic-1
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def check(root,lower,upper):
            if not root:
                return True
            val=root.val
            if val<=lower or val>=upper:
                return False
            if not check(root.left,lower,val):
                return False
            if not check(root.right,val,upper):
                return False
            return True
        lower=float('-inf')
        upper=float('+inf')
        return check(root,lower,upper)
# Logic-2
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        if not root:
            return True
        stack=[(root,float('-inf'),float('+inf'))]
        while stack:
            root,lower,upper=stack.pop()
            if not root:
                continue
            val=root.val
            if val<=lower or val>=upper:
                return False
            stack.append((root.right,val,upper))
            stack.append((root.left,lower,val))
        return True